### PR TITLE
docs: update cluster migration doc, closes #753

### LINKED ISSUE #753: [docs] Discussing alternative cluster migration strategies
We wanted to migrate from a kube-up-created 1.2.x cluster to a kops-managed 1.4.0 cluster with 0 downtime. After reading the docs in [upgrade_from_k8s_12.md](https://github.com/kubernetes/kops/blob/master/docs/upgrade_from_k8s_12.md), we decided to go with the following strategy to avoid the possibility of downtime. The following steps are taken from the writeup I did internally to document our approach.

1. Delegate cluster-level dns resolution to Route53 by adding appropriate NS records pointing `cluster.example.com` to Route53's Hosted Zone's nameservers.
2. Create the new cluster's configuration files with kops
    - `kops create cluster --cloud=aws --zones=us-east-1a,us-east-1b --admin-access=12.34.56.78/32 --dns-zone=cluster.example.com --kubernetes-version=1.4.0 --node-count=14 --node-size=c3.xlarge --master-zones=us-east-1a --master-size=m4.large --vpc=vpc-123abcdef --network-cidr=172.20.0.0/16 cluster.example.com` 
  - note that the `--network-cidr` is the cidr of the existing VPC.
  - Note that `kops` will propose re-naming the existing VPC but the change never occurs. We eventually manually changed it to the correct value.
3. Verify that the CIDR on each of the zone subnets does not overlap with an existing subnet's.
4. Verify the planned changes with `kops update cluster cluster.example.com`
5. Create the cluster with `kops update cluster cluster.example.com --yes`
6. Wait around for the cluster to fully come up and be available. `k get nodes` should return `(master + minions) = 15` available nodes.
7. (Optional) Create the Dashboard with `kubectl create -f https://rawgit.com/kubernetes/dashboard/master/src/deploy/kubernetes-dashboard.yaml`
8. Add the generated `nodes.cluster.example.com` security group to the resources that may need it (i.e. ElastiCache, RDS, etc)
9. Deploy the same resource configuration to the new cluster.
10. Wait patiently while everything happens.
11. Test that everything "just works" by hitting the ELB's dns entry directly; you should be able to use the site as expected.
12. To transition traffic from the old cluster to the new cluster, update the `CNAME` record for `example.com` to point to the new ELB's DNS name.
    - note that if you're proxying through cloudflare, changes are instantaneous because it's technically a reverse proxy and not a DNS change.
    - if not using cloudflare or another reverse proxy solution, you'll want to update you DNS record's TTL to a very low duration about 48 hours in advance of this change (and then change it back to the previous value once the shift has been finalized).
13. Rejoice.
14. Once traffic has shifted from the old cluster (relatively quickly), delete the old resources after confirming that traffic has stabilized and that no new errors are generated.

## Resources

This is a list of resource the old cluster is using that should be deleted. This may not be complete.
- autoscaling group (`kubernetes-minion-group-us-east-1a`)
- launch config (`kubernetes-minion-group-us-east-1a` and `kubernetes-minion-group-us-east-1aCopy`)
- all ec2 instances (`tag:KubernetesCluster : kubernetes`)
- all associated EBS volumes (most likely released when the instances are terminated)
- security groups (`tag:KubernetesCluster : kubernetes`)

## Recovery/Rollback

The only part of this procedure that should affect the users actively using the site is the DNS swap, which should be relatively instantaneous because we're using Cloudflare as a reverse proxy, not just as a nameserver. In the event that we need to swap back to the old cluster, we can replace the example.com CNAME record with the original ELB's DNS name.

---

I'm willing to write up a PR to add this to the document about migrations from previous clusters.

Are there any downsides we can surface around this approach? I believe kube-up also uses the route table, so we're unable to use this strategy for a cluster that requires `>25` nodes (since 50 is the artificial limitation on VPC Route Tables). If you're able to gradually transition traffic from one cluster to another (perhaps another load balancer to split traffic between clusters) you could slowly scale the new cluster up and the old cluster down to avoid hitting the 50-entry cap on the route table. Not an ideal situation, though.

### PR DESCRIPTION
**Pull request by Shrugs**
Updated the doc for upgrading and migrating clusters, with and without downtime.

Let me know if I could word things better or what.

cc @chrislovecnm

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kops/889)

### CODE REVIEW

**Comment by chrislovecnm** [comment]
Sorry that I have not reviewed this ... I have been fighting dragons

**Comment by Shrugs** [comment]
No worries!

File: docs/cluster_upgrades_and_migrations.md [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+At some point you will almost definitely want to upgrade the Kubernetes version of your cluster, or even migrate from a cluster managed/provisioned by another tool to one managed by `kops`. There are a few different ways to accomplish this depending on your existing cluster situation and any requirements for zero-downtime migrations.
+
+- Upgrade an existing `kube-up` managed cluster to one managed by `kops`
+    + [The simple method with downtime](#kube-up---kops-downtime)
+    + [The more complex method with zero-downtime](#kube-up---kops-sans-downtime)
+- [Upgrade a `kops` cluster from one Kubernetes version to another](upgrade.md)
+
+## `kube-up` -> `kops`, with downtime
+
+`kops` lets you upgrade an existing 1.2 cluster, installed using `kube-up`, to a cluster managed by `kops` running the latest kubernetes version (or the version of your choice).
```
Comment (chrislovecnm) [inline docs/cluster_upgrades_and_migrations.md]: > `kops` lets you upgrade an existing 1.2 cluster

`kops` lets you upgrade an existing 1.x cluster

File: docs/cluster_upgrades_and_migrations.md [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+### Other fixes
+
+* If you're using a manually created ELB, the auto-scaling groups change, so you will need to reconfigure your ELBs to include the new auto-scaling group(s).
+
+* It is recommended to delete old kubernetes system services that we imported (and replace them with newer versions):
+
+```
+kubectl delete rc -lk8s-app=kube-dns --namespace=kube-system
+
+kubectl delete rc -lk8s-app=elasticsearch-logging --namespace=kube-system
```
Comment (chrislovecnm) [inline docs/cluster_upgrades_and_migrations.md]: Should we add a note that those services may not be running?

File: docs/cluster_upgrades_and_migrations.md [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+You will also need to release the old ElasticIP manually.
+
+## `kube-up` -> `kops`, sans downtime
+
+### Overview
+
+This method provides zero-downtime when migrating a cluster from `kube-up` to `kops`. It does so by creating a logically separate `kops`-managed cluster in the existing `kube-up` VPC and then swapping the DNS entries (or your reverse proxy's upstream) to point to the new cluster's services.
+
+Limitations:
+- We're spinning up two identical clusters. Due to the default limit of 50 entries in a VPC's route table, if your cluster contains more than ~25 nodes, this strategy, as-is, will not work.
```
Comment (chrislovecnm) [inline docs/cluster_upgrades_and_migrations.md]: What about using cni networking?

File: docs/upgrade.md [hunk 4]
```diff
@@ -1,11 +1,10 @@
 ## Upgrading Kubernetes
 
-Upgrading Kubernetes is easy with kops.  The cluster spec contains a KubernetesVersion, so you
-can simply edit it with `kops edit`, and apply the updated configuration to your cluster.
+Upgrading Kubernetes is easy with kops.  The cluster spec contains a `KubernetesVersion`, so you can simply edit it with `kops edit`, and apply the updated configuration to your cluster.
```
Comment (chrislovecnm) [inline docs/upgrade.md]: should we add the fact that we can do binary for KubernetesVersion or a valid released version?

**Comment by chrislovecnm** [comment]
Bump

**Comment by chrislovecnm** [comment]
@Shrugs are you able to make the changes, so that we can get this in?

**Comment by Shrugs** [comment]
@chrislovecnm will tackle this today!

File: docs/upgrade.md (hunk 4 above)
Comment (Shrugs) [inline docs/upgrade.md]: What would specifying a binary look like? And do you mean that only valid released k8s versions are supported here?

**Comment by Shrugs** [comment]
@chrislovecnm can you take another look at the wording around https://github.com/kubernetes/kops/pull/889#discussion_r90382431 ?

**Comment by chrislovecnm** [comment]
@Shrugs can we get a rebase?

**Comment by chrislovecnm** [comment]
@Shrugs looks great, if we can get a rebase.

**Comment by Shrugs** [comment]
@chrislovecnm done!

**Comment by chrislovecnm** [comment]
Waiting on E2E

**Comment by chrislovecnm** [comment]
@k8s-bot test it

**Comment by chrislovecnm** [comment]
@k8s-bot aws e2e test this

**Comment by justinsb** [comment]
Thanks - I'm going to validate in https://github.com/kubernetes/kops/issues/1650, but not sure why we haven't merged this yet (?)

**Pull request merged**